clients/src/main/java/org/apache/kafka/common/requests/FetchResponse.java

```diff
@@ -86,6 +85,25 @@ public AbortedTransaction(long pid, long firstOffset) {
         }
 
         @Override
+        public boolean equals(Object o) {
+            if (this == o)
+                return true;
+            if (o == null || getClass() != o.getClass())
+                return false;
+
+            AbortedTransaction that = (AbortedTransaction) o;
+
+            return pid == that.pid && firstOffset == that.firstOffset;
+        }
+
+        @Override
+        public int hashCode() {
+            int result = (int) (pid ^ (pid >>> 32));
```

Comment:
I think there is tradeoff between code complexity and performance. It just feels weird/old-fashioned that we have to manually calculate hashCode for every request/response like this. I am OK with this if the choice between code complexity and performance is agreed upon by most people and we do this consistently across all classes (particularly request and response).

Ping @junrao and @hachikuji to comment on the tradeoff between code complexity and performance here.

Another question - it seems that we do not override `equals` and `hashCode` for other classes such as ProduceRequest. What is the motivation of overriding these two method for `FetchRequest`? If it is needed now, should we do it also for other classes? If it is not needed, maybe we should keep the code simple until it is needed?